src/asm-functions.cpp

```diff
@@ -287,80 +215,67 @@ bool ValidateUtf8StringSSE42(uint8_t const* src, std::size_t len) {
   return Utf8Helper::isValidUtf8(src, len);
 }
 
-bool doInitValidateUtf8String(uint8_t const* src, std::size_t limit) {
-#ifdef __AVX2__
-  if (assemblerFunctionsEnabled() && ::hasAVX2()) {
-    ValidateUtf8String = ValidateUtf8StringAVX;
-    return ValidateUtf8StringAVX(src, limit);
-  }
-#endif
-#if defined(__SSE4_2__)
-  if (assemblerFunctionsEnabled() && ::hasSSE42()) {
-    ValidateUtf8String = ValidateUtf8StringSSE42;
-  } else {
-    ValidateUtf8String = ::ValidateUtf8StringC;
-  }
-#elif defined(__aarch64__)
-  ValidateUtf8String = ValidateUtf8StringSSE42;
-#else
-  ValidateUtf8String = ::ValidateUtf8StringC;
 #endif
-  return (*ValidateUtf8String)(src, limit);
-}
-}  // namespace
-
-#else
-
-namespace {
-
-std::size_t doInitCopy(uint8_t* dst, uint8_t const* src, std::size_t limit) {
-  JSONStringCopy = ::JSONStringCopyC;
-  return ::JSONStringCopyC(dst, src, limit);
-}
+#ifdef __AVX2__
 
-std::size_t doInitCopyCheckUtf8(uint8_t* dst, uint8_t const* src,
-                                std::size_t limit) {
-  JSONStringCopyCheckUtf8 = ::JSONStringCopyCheckUtf8C;
-  return ::JSONStringCopyCheckUtf8C(dst, src, limit);
+bool hasAVX2() noexcept {
+  unsigned int eax, ebx, ecx, edx;
+  if (__get_cpuid(7, &eax, &ebx, &ecx, &edx)) {
+    if ((ebx & bit_AVX2) != 0) {
+      return true;
+    }
+  }
+  return false;
 }
 
-std::size_t doInitSkip(uint8_t const* src, std::size_t limit) {
-  JSONSkipWhiteSpace = ::JSONSkipWhiteSpaceC;
-  return JSONSkipWhiteSpace(src, limit);
+bool ValidateUtf8StringAVX(uint8_t const* src, std::size_t len) {
+  if (len >= 32) {
+    return validate_utf8_fast_avx(src, len);
+  }
+  return Utf8Helper::isValidUtf8(src, len);
 }
 
-bool doInitValidateUtf8String(uint8_t const* src, std::size_t limit) {
-  ValidateUtf8String = ::ValidateUtf8StringC;
-  return ValidateUtf8StringC(src, limit);
-}
+#endif
 
 }  // namespace
 
-#endif
-
-std::size_t (*JSONStringCopy)(uint8_t*, uint8_t const*,
-                              std::size_t) = ::doInitCopy;
+std::size_t (*JSONStringCopy)(uint8_t*, uint8_t const*, std::size_t) = nullptr;
```

Comment:
Why is every function pointer initialized to a `nullptr`?
That leaves room for someone using the function pointer before the initialize routine is called.
I suggest we initialize the function pointers to the portable C functions, so that the function pointers can never be `nullptr` and are always safe to use.

Reply:
Fixed now

Reply:
It was for check init called in tests

Reply:
You can see tests result before last commit

Reply:
Now it always correct.

Before static init we use c functions, after we use asm functions.

And the same time we can enable c/asm functions again